frontend/src/toolbar/elements/heatmapToolbarMenuLogic.ts

```diff
@@ -115,6 +116,21 @@ export const heatmapToolbarMenuLogic = kea<heatmapToolbarMenuLogicType>([
                 setSamplingFactor: (_, { samplingFactor }) => samplingFactor,
             },
         ],
+        elementMetrics: [
+            new Map<HTMLElement, { visible: boolean; rect: DOMRect }>(),
+            {
+                updateElementMetrics: (state, { element, visible, rect }) => {
+                    // we only change if visible or rect has changed
+                    const current = state.get(element)
+                    if (current?.visible === visible && current?.rect === rect) {
+                        return state
+                    }
+                    const newMap = new Map(state)
```

Comment:
Does it mean that you recreate a new map on each valid update? Shouldn't it be crazy expensive, though?

Reply:
we learn hard to reducers being immutable
so 🤞 😅